### PR TITLE
Inject custom headers defined during enroll

### PR DESCRIPTION
**Pull request by michalpristas**
## What does this PR do?

This PR takes custom headers defined during enroll and injects them into Elasticsearch config. 
Other output types do not pick these up. 
This is due to https://github.com/elastic/beats/issues/32993 which was introduced by V2 in a breaking manner. 

## Why is it important?

Every deployment with EA/APM enabled, that is using traffic filter will be broken on upgrade to 8.6.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)

### CODE REVIEW

File: pkg/component/component_test.go [hunk 1]
```diff
@@ -1588,14 +1589,173 @@ func TestToComponents(t *testing.T) {
...
+				"header-one": "val-1",
+			}},
+		},
+		{
+			Name:     "Headers injection not injecting kafka",
+			Platform: linuxAMD64Platform,
+			Policy: map[string]interface{}{
+				"outputs": map[string]interface{}{
+					"default": map[string]interface{}{
+						"type":    "kafka",
```
Comment (AndersonQ) [inline pkg/component/component_test.go:1707]: And logstash? Shouldn't we have one test for it as well?

File: pkg/component/component.go [hunk 2]
```diff
@@ -23,9 +23,15 @@ import (
 // GenerateMonitoringCfgFn is a function that can inject information into the model generation process.
 type GenerateMonitoringCfgFn func(map[string]interface{}, []Component, map[string]string) (map[string]interface{}, error)
 
+type HeadersProvider interface {
+	Headers() map[string]string
+}
```
Comment (AndersonQ) [inline pkg/component/component.go:28]: I wonder if this interface is really needed, a `map[string]string` is pretty standard for headers. Anyway, it isn't a blocker

**Review by AndersonQ — APPROVED** [review]

File: pkg/component/component.go (hunk 2 above)
Comment (michalpristas) [reply pkg/component/component.go:28]: i did this so i can pass agentinfo object but test with custom thingy. 
i dont want to pass headers as they are, but rather something which provides me up to date flags.

File: pkg/component/component.go [hunk 3]
```diff
@@ -470,6 +476,28 @@ func toIntermediate(policy map[string]interface{}, aliasMapping map[string]strin
 		if err != nil {
 			return nil, fmt.Errorf("invalid 'outputs.%s.log_level', %w", name, err)
 		}
+
+		// inject headers configured during enroll
+		if t == elasticsearchType && headers != nil {
```
Comment (blakerouse) [inline pkg/component/component.go:481]: Why is this elasticsearch specific?

Overall I am not super happy with injecting random headers, how do we know these headers apply only to an output? Do they apply to only one? What if there are 2 elasticsearch outputs, should they both get them or only one?

Comment (cmacknz) [reply pkg/component/component.go:481]: This is to support adding custom headers at enrolment time, a feature that was done with a config transformation rule prior to 8.6: https://github.com/elastic/elastic-agent/commit/69d8ae7bb3f7ec34a6f6d8b7a82f0d9bc1b86bd8

We removed this rule in 8.6 without replacing the underlying functionality and apparently broke some cloud deployments that rely on these enrolment time headers for traffic filtering.

I think the intent is that it should apply to all ES outputs.

**Pull request merged**